runtime/vm/createramclass.cpp

```diff
@@ -1584,15 +1594,26 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
 				if ((!superClassIsPublic && (packageID != superclass->packageID))
 					|| (superClassIsPublic && (J9_VISIBILITY_ALLOWED != checkModuleAccess(vmThread, vm, romClass, module, superclass->romClass, superclass->module, superclass->packageID, 0)))
 				) {
-					Trc_VM_CreateRAMClassFromROMClass_superclassNotVisible(vmThread, superclass, superclass->classLoader, classLoader);
-					setCurrentExceptionForBadClass(vmThread, superclassName, J9VMCONSTANTPOOL_JAVALANGILLEGALACCESSERROR);
-					return FALSE;
+					if (J9CLASS_IS_EXEMPT_FROM_VALIDATION(superclass)) {
+						/* this class is exempt like its superclass */
+						isExemptFromValidation = TRUE;
+					} else {
+						Trc_VM_CreateRAMClassFromROMClass_superclassNotVisible(vmThread, superclass, superclass->classLoader, classLoader);
+						setCurrentExceptionForBadClass(vmThread, superclassName, J9VMCONSTANTPOOL_JAVALANGILLEGALACCESSERROR);
+						return FALSE;
+					}
 				}
 			}
 
+			if (J9_ARE_ANY_BITS_SET(options, J9_FINDCLASS_FLAG_NAME_IS_INVALID) && !isExemptFromValidation) {
```

Comment:
Should this be setting the bit and letting the caller deal with the exception? The exception below is a much better one than is being thrown in the new calling code.

Reply:
When you say 'below', do you mean the one in `clsldr.cpp`? Just returning `FALSE` here with no exception seems to work (and is reasonable given the caller has claimed responsibility for dealing with the invalid name).

Reply:
"below" I mean the special exception setter call which has good detail info. The one being thrown outside is bare (no message). I'd move the good call out to the caller and check there (which you indicate works).

Reply:
I spoke too soon: calling `ClassLoader.loadClass()` with a `/` name referring to an already loaded class yields a `LinkageError` instead of the more specific `NoClassDefFoundError`. I'll see if there's a straight-forward fix for that.

Reply:
Strictly speaking, that's fine - you are throwing a LinkageError. Unless there's a test querying the exact class of the exception, I think that should be fine.

Comment welcome from @DanHeidinga @pshipton

Reply:
While GAC is likely technically correct, the javadoc for loadClass() only mentions `ClassNotFoundException` so I expect there's a bunch of code that knows to map `NoClassDefFoundError` -> `ClassNotFoundException` and may not handle LinkageError correctly

Reply:
OpenJ9 has at least one test that expects `NoClassDefFoundError` and considers the more general `LinkageError` a failure: that expected behavior is consistent with the RI.

Reply:
I only know of the reverse mapping (`ClassNotFoundException` -> `NoClassDefFoundError`) that used to be involved in the compiled code for expressions of the form `Whatever.class` before class constants were a thing in class files: the compiler knew about the mentioned class and so generates code to catch `ClassNotFoundException` (a checked exception) thrown by `Class.forName()` and map it to `NoClassDefFoundError`.